src/pandablocks_ioc/_hdf_ioc.py

```diff
@@ -522,6 +540,41 @@ def _parameter_validate(self, record: RecordWrapper, new_val) -> bool:
         return True
 
     async def _update_full_file_path(self, new_val) -> None:
+        full_path_as_list = os.path.normpath(new_val).split(os.sep)
+        drive = full_path_as_list[0]
+        new_path_as_list = full_path_as_list[1:]
+        create_dir_depth = self._create_directory_record.get()
+
+        dirs_to_create = 0
+        if create_dir_depth < 0:
+            if len(new_path_as_list) < abs(create_dir_depth):
+                dirs_to_create = len(new_path_as_list)
+            else:
+                dirs_to_create = abs(create_dir_depth)
+        elif create_dir_depth > 0:
+            if len(new_path_as_list) < create_dir_depth:
+                dirs_to_create = 0
+            else:
+                dirs_to_create = len(new_path_as_list) - create_dir_depth
+
+        if dirs_to_create > 0:
+            if len(full_path_as_list[:-2]) == 1:
+                # Insert a blank element in index #1 if at root dir.
+                # Ex. on linux, [''] becomes ['', ''] which joins to '/'
+                # On windows ['C:'] becomes ['C:', ''] which joins to 'C:\\'
+                full_path_as_list.insert(1, "")
+            if os.path.isdir(os.sep.join(full_path_as_list[:-dirs_to_create])) and not os.path.exists(new_val):
+                try:
+                    os.makedirs(new_val, exist_ok=True)
+                    logging.info(f"Created directory {new_val}.")
+                except Exception as e:
+                    logging.error(f"Failed to create directory {new_val}! Error: {str(e)}")
+
+        if os.path.exists(new_val) and os.access(new_val, os.W_OK):
+            self._directory_exists_record.set(1)
+        else:
+            self._directory_exists_record.set(0)
```

Comment:
I think this code can be simplified by making use of `pathlib.Path` and its platform-agnostic handling of directory paths. For example:
```
def _update_full_file_path(self, new_val):
    # This assumes new_val is a full file path relative to
    # the current directory e.g. dir1/dir2/file.hdf
    new_path = Path(new_val)

    val = self._create_directory_record.get()
    if val < 0:
        create_dir_count = abs(val)
    else:
        create_dir_count = len(new_path.parent.parts) - val

    # Iterate from top-most directory downwards, creating directories as
    # necessary up to maximum amount
    for p in reversed(new_path.parents):
        if not p.exists() and abs(create_dir_count) > 0:
            p.mkdir()
            create_dir_count -= 1

    if new_path.parent.is_dir() and os.access(new_path.parent, os.W_OK):
        self._directory_exists_record.set(1)
    else:
        self._directory_exists_record.set(0)
```
I haven't rigorously tested this, as I don't entirely understand the reasoning behind positive values of CreateDirectory. I don't understand why you would ever want to partially create a directory tree, or have such fine-grained control as to whether the directory is created or not, so these semantics are probably not represented correctly in this example code.

Reply:
I think the reason for not auto-creating a directory tree is because of the expectation that part of the tree is expected to already exist with certain ownership permissions, and having the IOC process touch these directories may lead to permissions/ownership potentially changing in unwanted ways. From my view auto-creating all dirs in the tree is fine, but I think we should have the same behavior as AD for portability sake in things like writers in ophyd.